streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -623,6 +624,16 @@ private void closeGlobalStreamThread() {
     public synchronized boolean close(final long timeout, final TimeUnit timeUnit) {
         log.debug("{} Stopping Kafka Stream process.", logPrefix);
 
+        // do not close during state changes, e.g., there are corner cases when all
+        // threads die and the final thread to die calls onChange() is called to set the instances state.
+        // Within the instance callback, the user might call close() on the instance, which will attempt to join()
+        // the thread that started. So we have a circular loop (deadlock). We cannot stop the user from calling close()
+        // however we can prevent it from occurring.
+        if (inTheMiddleOfStateChange) {
```

Comment:
One concern with this is that the other `close` method is `void` return and also the comment says:
```
/**
     * Shutdown this {@code KafkaStreams} instance by signaling all the threads to stop, and then wait for them to join.
     * This will block until all threads have stopped.
     */
```
So, a user would expect the instance to be closed once that call returns. However with this change, that may not be the case.